### PR TITLE
View for event participants

### PR DESCRIPTION
**Pull request by nikhilgupta1211**

### CODE REVIEW

**Comment by nikhilgupta1211** [comment]
`events#show` view

image

`events#participants` view

image

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -4,6 +4,12 @@ class EventsController < InheritedResources::Base
   skip_load_and_authorize_resource only: [:index, :show]
   before_filter :set_types
 
+  def participants
+    @breadcrumbs = [label: 'events', url: events_path]
+    @breadcrumbs << { label: @event.name, url: event_path(@event) }
+    @breadcrumbs << { label: 'participants' }
```
Comment (bgeuken) [inline app/controllers/events_controller.rb]: Since this is static data, I would prefer to assign all data to ``` @breadcrumbs ``` at once.

File: app/views/events/participants.html.haml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  %br
+  %table.table
+    %tr
+      %th #
+      %th Nickname
+      %th Email
+      %th State
+      %th Role
+      %th Change role to event organizer
+    - @event.requests.each_with_index do |e,i|
```
Comment (bgeuken) [inline app/views/events/participants.html.haml]: Please use more speaking variable names for iterators that go over more than one line of code, eg index and event

Comment (nikhilgupta1211) [inline app/views/events/participants.html.haml]: ok :)

File: app/controllers/events_controller.rb (hunk 1 above)
Comment (nikhilgupta1211) [inline app/controllers/events_controller.rb]: yes ok

**Comment by nikhilgupta1211** [comment]
image

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -4,6 +4,10 @@ class EventsController < InheritedResources::Base
   skip_load_and_authorize_resource only: [:index, :show]
   before_filter :set_types
 
+  def participants
+    @breadcrumbs = [{ label: 'events', url: events_path }, { label: @event.name, url: event_path(@event) }, { label: 'participants' }]
```
Comment (bgeuken) [inline app/controllers/events_controller.rb]: Of course a matter of taste. I would prefer this format:

```ruby
@breadcrumbs = [
  { label: 'events', url: events_path },
  { label: @event.name, url: event_path(@event) },
  { label: 'participants' }
] 
```

@ChrisBr @nikhilgupta1211  What do you say?

File: app/views/events/participants.html.haml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  %br
+  %table.table
+    %tr
+      %th #
+      %th Nickname
+      %th Email
+      %th State
+      %th Role
+      %th Change role to event organizer
+    - @event.requests.map(&:user).uniq.each_with_index do |user,index|
```
Comment (ChrisBr) [inline app/views/events/participants.html.haml]: You could use ``with_index(1)`` so you don't need to add +1 in the loop.
https://apidock.com/ruby/Enumerator/with_index

Also I don't really like the query and mapping etc. I would prefer to do that in the view in a proper way.

```
User.where(request: @event).includes(:request).distinct.
```

This would also trigger several n+1 queries
http://guides.rubyonrails.org/active_record_querying.html#eager-loading-associations

File: app/controllers/events_controller.rb [hunk 5]
```diff
@@ -4,6 +4,14 @@ class EventsController < InheritedResources::Base
   skip_load_and_authorize_resource only: [:index, :show]
   before_filter :set_types
 
+  def participants
```
Comment (ChrisBr) [inline app/controllers/events_controller.rb]: Hm, @bgeuken what do you think about introducing controller namespaces as I don't like introducing participants here.

http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/

File: app/helpers/events_helper.rb [hunk 6]
```diff
@@ -16,4 +16,28 @@ def users_for_event(state)
     req = req.where(state: state) if state != 'all'
     user_email = req.map { |e| e.user.email }.uniq
   end
+
+  def check_states(user)
+    requests = user.requests.where(event: @event)
```
Comment (ChrisBr) [inline app/helpers/events_helper.rb]: Is there a way to not trigger another db query again?

File: app/views/events/participants.html.haml (hunk 4 above)
Comment (bgeuken) [inline app/views/events/participants.html.haml]: Right, the query needs some love. I guess a join would make sense here

File: app/controllers/events_controller.rb (hunk 5 above)
Comment (bgeuken) [inline app/controllers/events_controller.rb]: Yes, why not. Or, if the PR get's to complex we could refactor this once this got merged.

File: app/views/events/participants.html.haml (hunk 4 above)
Comment (nikhilgupta1211) [inline app/views/events/participants.html.haml]: @ChrisBr thanks you :), I didn't knew about the eager loading associations.

Before it was taking 12 queries now it is taking only 1 after changing it to :
`User.includes(:requests, :profile).where(requests: {event: @event})`

File: app/controllers/events_controller.rb (hunk 5 above)
Comment (nikhilgupta1211) [inline app/controllers/events_controller.rb]: Can you please tell that should we do the refactoring in this PR only ?

File: app/views/events/participants.html.haml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  %br
+  %table.table
+    %tr
+      %th #
+      %th Nickname
+      %th Email
+      %th State
+      %th Role
+      %th Change role to event organizer
+    - User.includes(:requests, :profile).where(requests: {event: @event}).each.with_index(1) do |user,index|
```
Comment (ChrisBr) [inline app/views/events/participants.html.haml]: While thinking about it: What about having a has_many through association in Request
http://guides.rubyonrails.org/association_basics.html#the-has-many-through-association

and then you can do in the controller 

```
@users = @event.users.distinct
```

Caution: Untested and please check if the distinct is necessary

File: app/helpers/events_helper.rb [hunk 8]
```diff
@@ -16,4 +16,27 @@ def users_for_event(state)
     req = req.where(state: state) if state != 'all'
     user_email = req.map { |e| e.user.email }.uniq
   end
+
+  def check_states(requests)
+    if requests.map { |r| r.event = @event }.count > 1
```
Comment (bgeuken) [inline app/helpers/events_helper.rb]: I would prefer to do this in sql, like joining the events and then limit the result to 2.

Comment (ChrisBr) [inline app/helpers/events_helper.rb]: Hm, this is not nice! You could do sth like this ```requests.where(event: @event).count``` but that would still trigger n + 1 queries. @bgeuken any better solution? We could also iterate over the requests in the view but I think this is strange because the table is about users and not about requests and then we would need to do always request.user.foo to access attributes ...

Comment (nikhilgupta1211) [inline app/helpers/events_helper.rb]: I checked and requests.where(event: @event).count do not trigger n+1 queries

File: app/controllers/events_controller.rb (hunk 5 above)
Comment (bgeuken) [inline app/controllers/events_controller.rb]: Keep it for now. It can be refactored on a separate PR.

Comment (nikhilgupta1211) [inline app/controllers/events_controller.rb]: ok thanks :)

File: app/views/events/participants.html.haml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  %br
+  %table.table
+    %tr
+      %th #
+      %th Nickname
+      %th Email
+      %th State
+      %th Role
+      %th Change role to event organizer
+    - @event.users.includes(:profile).distinct.each.with_index(1) do |user,index|
```
Comment (bgeuken) [inline app/views/events/participants.html.haml]: Much nicer :+1:

Comment (bgeuken) [inline app/views/events/participants.html.haml]: But since within the loop we still query the requests, you might want to include the requests here.

Comment (nikhilgupta1211) [inline app/views/events/participants.html.haml]: No we are not quering for the requests here as we are just sending all the requests of a user in the method.

Comment (bgeuken) [inline app/views/events/participants.html.haml]: But [user requests](https://github.com/openSUSE/travel-support-program/pull/80/files/98500327a95b33f18b94f0ed1915d081e1b02add#diff-018ebd24bb84fbd780e127372855d5c5R20) would result in another db query, right? Including the requests before would cause rails to preload them and thus reduce the number of queries.

Comment (nikhilgupta1211) [inline app/views/events/participants.html.haml]: I tried but it is still taking the same number of queries

File: spec/features/events_spec.rb [hunk 10]
```diff
@@ -72,8 +74,13 @@
...
     click_link "Death Star's destruction celebration"
     page.should have_link('Email')
+    page.should have_link('Participants')
 
     click_link 'Email'
     page.should have_content "Emails for Death Star's destruction celebration"
+
+    visit event_path(events(:party))
+    click_link 'Participants'
+    page.should have_content "Participants of Death Star's destruction celebration"
```
Comment (bgeuken) [inline spec/features/events_spec.rb]: Please also test at least one or two rows here, like are all users listed here (eg. count the rows)? and check one or two users in detail (eg. do they have correct state, role, etc.)

Comment (nikhilgupta1211) [inline spec/features/events_spec.rb]: ok :)

File: app/helpers/events_helper.rb [hunk 11]
```diff
@@ -16,4 +16,27 @@ def users_for_event(state)
...
+  def state_label(state)
+    case state
+    when 'submitted'
+      content_tag(:span, state, class: 'label label-primary')
+    when 'canceled'
+      content_tag(:span, state, class: 'label label-danger')
+    when 'incomplete'
+      content_tag(:span, state, class: 'label label-warning')
+    when 'accepted'
+      content_tag(:span, state, class: 'label label-info')
```
Comment (ChrisBr) [inline app/helpers/events_helper.rb]: I guess we're lacking the Reimbursable states, don't we?

File: app/models/event.rb [hunk 12]
```diff
@@ -13,6 +13,8 @@ class Event < ActiveRecord::Base
   has_many :shipments, inverse_of: :event, dependent: :restrict_with_exception
   # Mails for an event
   has_many :event_emails
+  # Users for an event
+  has_many :users, through: :requests
```
Comment (nikhilgupta1211) [inline app/models/event.rb]: yes ok

**Comment by nikhilgupta1211** [comment]
image
@ChrisBr @bgeuken I have updated the PR please have a look at it :)

**Comment by bgeuken** [comment]
LGTM

**Comment by bgeuken** [comment]
UI looks good to me as well

File: app/controllers/events_controller.rb [hunk 13]
```diff
@@ -4,6 +4,15 @@ class EventsController < InheritedResources::Base
...
+    @breadcrumbs = [
+      { label: 'events', url: events_path },
+      { label: @event.name, url: event_path(@event) },
+      { label: 'participants' }
+    ]
+    @users = @event.users
```
Comment (ChrisBr) [inline app/controllers/events_controller.rb]: Can you test please that it only returns users of requests where the user has abilities

Comment (nikhilgupta1211) [inline app/controllers/events_controller.rb]: yes ok

File: app/models/event.rb [hunk 14]
```diff
@@ -13,8 +13,10 @@ class Event < ActiveRecord::Base
   has_many :shipments, inverse_of: :event, dependent: :restrict_with_exception
   # Mails for an event
   has_many :event_emails
+  # Travel Sponsorships for an event
+  has_many :travel_sponsorships
   # Users for an event
-  has_many :users, through: :requests
+  has_many :users, -> { distinct }, through: :travel_sponsorships
```
Comment (bgeuken) [inline app/models/event.rb]: I think it would be nicer here to keep the "through: :requests" and add a scope for the travel sponsorship.

That's more flexible and allows us to query for other request types in the future.

Comment (bgeuken) [inline app/models/event.rb]: And move the destinct to the scope as well.

Comment (nikhilgupta1211) [inline app/models/event.rb]: sorry but can you please tell how to define a scope for the requests in the has_many : users statement

Comment (nikhilgupta1211) [inline app/models/event.rb]: I have made some changes can you please tell if they are ok

Comment (ChrisBr) [inline app/models/event.rb]: http://guides.rubyonrails.org/active_record_querying.html#scopes

File: app/models/event.rb [hunk 15]
```diff
@@ -14,7 +14,12 @@ class Event < ActiveRecord::Base
   # Mails for an event
   has_many :event_emails
   # Users for an event
-  has_many :users, through: :requests
+  has_many :users, -> { distinct }, through: :requests do
+    # Users with travel requests
+    def travel
+      where("requests.type == 'TravelSponsorship'")
```
Comment (ChrisBr) [inline app/models/event.rb]: Please use the ruby hash syntax instead!

Comment (ChrisBr) [inline app/models/event.rb]: Anyway, not sure if this is what @bgeuken meant ...

File: spec/features/events_spec.rb (hunk 10 above)
Comment (ChrisBr) [inline spec/features/events_spec.rb]: And also that the user is not able to see the row if he doesn't have the ability.

File: app/controllers/events_controller.rb [hunk 16]
```diff
@@ -4,6 +4,15 @@ class EventsController < InheritedResources::Base
   skip_load_and_authorize_resource only: [:index, :show]
   before_filter :set_types
 
+  def participants
+    @breadcrumbs = [
+      { label: 'events', url: events_path },
+      { label: @event.name, url: event_path(@event) },
+      { label: 'participants' }
+    ]
+    @users = @event.users.travel
```
Comment (ChrisBr) [inline app/controllers/events_controller.rb]: You're still not using accessible_by(current_ability), this can cause some privacy issues if you access data without checking the ability.

Comment (nikhilgupta1211) [inline app/controllers/events_controller.rb]: Actually, accessible_by checks if the current ability can fetch that record and if we use it here https://github.com/openSUSE/travel-support-program/pull/80/files#diff-3a4e95c1bc41d3110d2f9ac10064e79bR13 it will not be able to fetch any records because tsp members have no extra permission to view a User record.

A tsp member can only read the User Profile of a user https://github.com/openSUSE/travel-support-program/blob/master/app/models/ability.rb#L137

So we can define, can :read, Request, type: "TravelSponsorship" in the tsp role and then fetch the requests using accessible_by and then fetch the users from it, but I think it is better to use users only with the requests type as TravelSponsorship as a tsp member can view all these requests.

File: spec/features/events_spec.rb (hunk 10 above)
Comment (nikhilgupta1211) [inline spec/features/events_spec.rb]: yes ok :)

File: app/controllers/events_controller.rb (hunk 16 above)
Comment (ChrisBr) [inline app/controllers/events_controller.rb]: I don't think that this is a good idea as it would move the ability logic out of the ability file ...

Comment (nikhilgupta1211) [inline app/controllers/events_controller.rb]: But we are just using the travel request scope here so why it is moving out the ability logic ?

File: app/models/event.rb [hunk 17]
```diff
@@ -14,7 +14,12 @@ class Event < ActiveRecord::Base
   # Mails for an event
   has_many :event_emails
   # Users for an event
-  has_many :users, through: :requests
+  has_many :users, -> { distinct }, through: :requests do
+    # Users with travel requests
+    def travel
+      where('type = ?', 'TravelSponsorship')
```
Comment (ChrisBr) [inline app/models/event.rb]: Please use ruby hash syntax

Comment (nikhilgupta1211) [inline app/models/event.rb]: I tried using the ruby hash syntax but it is not working

Comment (ChrisBr) [inline app/models/event.rb]: What did you try? Sth like this should work:

```
where(requests: { type: 'TravelSponsorship' } )
```

Comment (nikhilgupta1211) [inline app/models/event.rb]: Sorry actually I tried this `where(type: 'TravelSponsorship')` .

Thanks :)

File: app/models/event.rb [hunk 18]
```diff
@@ -14,7 +14,12 @@ class Event < ActiveRecord::Base
   # Mails for an event
   has_many :event_emails
   # Users for an event
-  has_many :users, through: :requests
+  has_many :users, -> { distinct }, through: :requests do
+    # Users with travel requests
+    def travel
+      where(requests: { type: 'TravelSponsorship' } )
+    end
+  end
```
Comment (bgeuken) [inline app/models/event.rb]: :+1:

File: spec/features/events_spec.rb (hunk 10 above)
Comment (bgeuken) [inline spec/features/events_spec.rb]: @nikhilgupta1211 The permission check is still missing, right?

Comment (nikhilgupta1211) [inline spec/features/events_spec.rb]: @bgeuken sorry didn't get you

Comment (bgeuken) [inline spec/features/events_spec.rb]: The test Christian asked for. To test that a user can not see things he is not entitled to

File: spec/features/events_spec.rb [hunk 19]
```diff
@@ -72,8 +74,21 @@
...
+    page.all('.table tr').count.should == 7
+
+    page.should_not have_content 'john.skywalker@rebel-alliance.org' # User with a shipment request
```
Comment (nikhilgupta1211) [inline spec/features/events_spec.rb]: @bgeuken I have added a shipment request and defined this so that a tsp member is not able to see the user of a shipment request.

File: spec/features/events_spec.rb (hunk 10 above)
Comment (nikhilgupta1211) [inline spec/features/events_spec.rb]: @bgeuken https://github.com/openSUSE/travel-support-program/pull/80/files#diff-f20a1b372768b0b5eebe83bc6aa154f8R87 I am checking here that a tsp member is not able to see a user of a shipment request

File: app/controllers/events_controller.rb (hunk 16 above)
Comment (ChrisBr) [inline app/controllers/events_controller.rb]: Would it be possible to change the association to user_profile and use the user profile here instead? I think that would make it way more clear!

File: app/models/event.rb [hunk 20]
```diff
@@ -13,7 +13,13 @@ class Event < ActiveRecord::Base
   has_many :shipments, inverse_of: :event, dependent: :restrict_with_exception
   # Mails for an event
   has_many :event_emails
-
+  # Users for an event
+  has_many :users, -> { distinct }, through: :requests do
+    # Users with travel requests
+    def travel
```
Comment (ChrisBr) [inline app/models/event.rb]: This scope belongs to ``requests`` association I would say

File: app/views/events/show.html.haml [hunk 21]
```diff
@@ -13,6 +13,8 @@
         = link_to t(:new_travel), new_travel_sponsorship_path(:event_id => resource), :class => 'btn btn-primary'
       - if resource.accepting_shipments?
         = link_to t(:new_shipment), new_shipment_path(:event_id => resource), :class => 'btn btn-primary'
+      - if user_signed_in? && can?(:participants, resource)
```
Comment (ChrisBr) [inline app/views/events/show.html.haml]: Is this ability already defined?

**Comment by nikhilgupta1211** [comment]
@ChrisBr I have rebased it :)

**Comment by bgeuken** [comment]
@nikhilgupta1211 The first 2 commits are also in #90, right?

**Comment by nikhilgupta1211** [comment]
@bgeuken Yes that branch has to be rebased after we merge this :)
